File: sport-log-types/src/types/action.rs
use chrono::{DateTime, Utc};
use derive_deftly::Deftly;
#[cfg(feature = "db")]
use diesel::{deserialize::FromSqlRow, expression::AsExpression, prelude::*, sql_types::BigInt};
#[cfg(feature = "db")]
use diesel_derive_enum::DbEnum;
use serde::{Deserialize, Serialize};

#[cfg(feature = "db")]
use crate::{
    Platform, User,
    schema::{action, action_event, action_provider, action_rule},
};
use crate::{PlatformId, UserId, types::IdString};

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq, Deftly)]
#[derive_deftly(IdString)]
#[serde(try_from = "IdString", into = "IdString")]
#[cfg_attr(
    feature = "db",
    derive(Hash, FromSqlRow, AsExpression),
    derive_deftly(IntoPgBigInt, FromPgBigInt),
    diesel(sql_type = diesel::sql_types::BigInt)
)]
pub struct ActionProviderId(pub i64);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(
    feature = "db",
    derive(
        Insertable,
        Associations,
        Identifiable,
        Queryable,
        Selectable,
        AsChangeset,
    ),
    diesel(table_name = action_provider, belongs_to(Platform))
)]
pub struct ActionProvider {
    pub id: ActionProviderId,
    pub name: String,
    pub password: String,

    pub platform_id: PlatformId,
    #[cfg_attr(feature = "db", diesel(treat_none_as_null = true))]
    pub description: Option<String>,
    pub deleted: bool,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq, Deftly)]
#[derive_deftly(IdString)]
#[serde(try_from = "IdString", into = "IdString")]
#[cfg_attr(
    feature = "db",
    derive(Hash, FromSqlRow, AsExpression),
    derive_deftly(IntoPgBigInt, FromPgBigInt),
    diesel(sql_type = diesel::sql_types::BigInt)
)]
pub struct ActionId(pub i64);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(
    feature = "db",
    derive(
        Insertable,
        Associations,
        Identifiable,
        Queryable,
        Selectable,
        AsChangeset,
    ),
    diesel(table_name = action, belongs_to(ActionProvider))
)]
pub struct Action {
    pub id: ActionId,
    pub name: String,
    pub action_provider_id: ActionProviderId,
    #[cfg_attr(feature = "db", diesel(treat_none_as_null = true))]
    pub description: Option<String>,
    pub create_before: i32,
    pub delete_after: i32,
    pub deleted: bool,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(
    feature = "db",
    derive(DbEnum),
    ExistingTypePath = "crate::schema::sql_types::Weekday"
)]

pub enum Weekday {
    Monday,
    Tuesday,
    Wednesday,
    Thursday,
    Friday,
    Saturday,
    Sunday,
}

impl Weekday {
    pub fn to_u32(self) -> u32 {
        match self {
            Weekday::Monday => 0,
            Weekday::Tuesday => 1,
            Weekday::Wednesday => 2,
            Weekday::Thursday => 3,
            Weekday::Friday => 4,
            Weekday::Saturday => 5,
            Weekday::Sunday => 6,
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq, Deftly)]
#[derive_deftly(IdString)]
#[serde(try_from = "IdString", into = "IdString")]
#[cfg_attr(
    feature = "db",
    derive(Hash, FromSqlRow, AsExpression),
    derive_deftly(IntoPgBigInt, FromPgBigInt),
    diesel(sql_type = BigInt)
)]
pub struct ActionRuleId(pub i64);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(
    feature = "db",
    derive(
        Insertable,
        Associations,
        Identifiable,
        Queryable,
        Selectable,
        AsChangeset,
    ),
    diesel(table_name = action_rule, belongs_to(User), belongs_to(Action))
)]
pub struct ActionRule {
    pub id: ActionRuleId,
    pub user_id: UserId,
    pub action_id: ActionId,
    pub weekday: Weekday,
    pub time: DateTime<Utc>,
    #[cfg_attr(feature = "db", diesel(treat_none_as_null = true))]
    pub arguments: Option<String>,
    pub enabled: bool,
    pub deleted: bool,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq, Deftly)]
#[derive_deftly(IdString)]
#[serde(try_from = "IdString", into = "IdString")]
#[cfg_attr(
    feature = "db",
    derive(Hash, FromSqlRow, AsExpression),
    derive_deftly(IntoPgBigInt, FromPgBigInt),
    diesel(sql_type = BigInt)
)]
pub struct ActionEventId(pub i64);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(
    feature = "db",
    derive(
        Insertable,
        Associations,
        Identifiable,
        Queryable,
        Selectable,
        AsChangeset,
    ),
    diesel(table_name = action_event, belongs_to(User), belongs_to(Action))
)]
pub struct ActionEvent {
    pub id: ActionEventId,
    pub user_id: UserId,
    pub action_id: ActionId,
    pub datetime: DateTime<Utc>,
    #[cfg_attr(feature = "db", diesel(treat_none_as_null = true))]
    pub arguments: Option<String>,
    pub enabled: bool,
    pub deleted: bool,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "db", derive(Queryable))]
pub struct CreatableActionRule {
    pub action_rule_id: ActionRuleId,
    pub user_id: UserId,
    pub action_id: ActionId,
    pub weekday: Weekday,
    pub time: DateTime<Utc>,
    #[cfg_attr(feature = "db", diesel(treat_none_as_null = true))]
    pub arguments: Option<String>,
    pub create_before: i32,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "db", derive(Queryable))]
pub struct ExecutableActionEvent {
    pub action_event_id: ActionEventId,
    pub action_name: String,
    pub datetime: DateTime<Utc>,
    #[cfg_attr(feature = "db", diesel(treat_none_as_null = true))]
    pub arguments: Option<String>,
    pub user_id: UserId,
    #[cfg_attr(feature = "db", diesel(treat_none_as_null = true))]
    pub username: Option<String>,
    #[cfg_attr(feature = "db", diesel(treat_none_as_null = true))]
    pub password: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "db", derive(Queryable))]
pub struct DeletableActionEvent {
    pub action_event_id: ActionEventId,
    pub datetime: DateTime<Utc>,
    pub delete_after: i32,
}

File: sport-log-client/frontend/lib/database/tables/action.dart

import 'package:sport_log/database/table.dart';
import 'package:sport_log/models/action/all.dart';

class ActionTable extends Table<Action> {
  @override DbSerializer<Action> get serde => DbActionSerializer();
  @override String get setupSql => '''
create table action (
    id integer primary key,
    name text not null check (length(name) >= 2),
    action_provider_id integer not null references action_provider(id) on delete cascade,
    description text,
    create_before integer not null check (create_before >= 0), -- hours
    delete_after integer not null check (delete_after >= 0), --hours
    last_change text not null default (datetime('now')),
    deleted integer not null default 0 check (deleted in (0, 1)),
    unique (action_provider_id, name, deleted)
);
  ''';
  @override String get tableName => 'action';
}

class ActionEventTable extends Table<ActionEvent> {
  @override DbSerializer<ActionEvent> get serde => DbActionEventSerializer();
  @override String get setupSql => '''
create table action_event (
    id integer primary key,
    user_id integer not null,
    action_id integer not null references action(id) on delete cascade,
    datetime text not null,
    enabled integer not null check(enabled in (0, 1)),
    last_change text not null default (datetime('now')),
    deleted integer not null default 0 check (deleted in (0, 1)),
    unique (user_id, action_id, datetime, deleted)
);
  ''';
  @override String get tableName => 'action_event';
}

class ActionRuleTable extends Table<ActionRule> {
  @override DbSerializer<ActionRule> get serde => DbActionRuleSerializer();
  @override String get setupSql => '''
create table action_rule (
    id integer primary key,
    user_id integer not null,
    action_id bigint not null references action(id) on delete cascade,
    weekday integer not null check(weekday between 0 and 6), 
    time text not null,
    enabled integer not null check(enabled in (0, 1)),
    last_change text not null default (datetime('now')),
    deleted integer not null default 0 check (deleted in (0, 1)),
    unique (user_id, action_id, weekday, time, deleted)
);
  ''';
  @override String get tableName => 'action_rule';
}

class ActionProviderTable extends Table<ActionProvider> {
  @override DbSerializer<ActionProvider> get serde => DbActionProviderSerializer();
  @override String get setupSql => '''
create table action_provider (
    id integer primary key,
    name text not null check(length(name) >= 2),
    password text not null check(length(password) <= 96),
    platform_id integer not null references platform(id) on delete cascade,
    description text,
    last_change text not null default (datetime('now')),
    deleted integer not null default 0 check (deleted in (0, 1)),
    unique (name, deleted)
);
  ''';
  @override String get tableName => 'action_provider';
}